File: c++/sparseGraph.cpp
//this file is used to create a graph and realize breadth first search
//as well as depth first search
//in adjacent list
#include <iostream>
#include <stdio.h>
#include <string>
#include<vector>
#include<stack>
#include<queue>
using namespace std;
class SparseGraph{
public:
	int n, m;//n nodes, m edges
	bool directed;//directed or non-directed
	vector<vector<int> > g;//2D vector

	//construct function
	SparseGraph(int n, bool directed){
		this->n = n;
		this->directed = directed;
		for(int i = 0; i < n; i++){
			g.push_back(vector<int>());
		}
	}

	//deconstruct function
	~SparseGraph(){}

	//get node number
	int getV(){
		return n;
	}

	//get edge number
	int getE(){
		return m;
	}

	//add edge between two nodes
	void addEdge(int v, int w){
		g[v].push_back(w);//v -> w
		if(!directed){
			g[w].push_back(v);
		}//if non-directed, w -> v
		m++;//add an edge from v to w
	}

	//check whether has an edge
	bool hasEdge(int v, int w){
		for(int i = 0; i < g[v].size(); i++){
			if(g[v][i] == w) return true;
		}
		return false;
	}

	void print(){
		for(int i = 0; i < g.size(); i++){
			cout<<"edge "<<i<<" : ";
			for(int j = 0; j < g[i].size(); j++){
				cout<<g[i][j]<<" ";
			}
			cout<<endl;
		}
	}

	void dfs(int v){
		vector<bool> visited(n,false);
		stack<int> s;
		s.push(v);
		while(!s.empty()){
			int tmp = s.top();
			if(!visited[tmp]){ //if node isn't visited
				cout<<tmp<<" ";
			}
			visited[tmp] = true;
			s.pop();
			int size = g[tmp].size();//push each element non-visited connected with tmp
			for(int i = 0; i < size; i++){
				int b = g[tmp][i];
				if(!visited[b]) s.push(b);
			}
		}
		cout<<endl;
	}

	void bfs(int v){
		vector<bool> visited(n,false);
		queue<int> que;
		que.push(v);
		while(!que.empty()){
			int tmp = que.front();
			if(!visited[tmp]) cout<<tmp<<" ";
			visited[tmp] = true;
			que.pop();
			int size = g[tmp].size();
			for(int i = 0; i < size; i++){
				int b = g[tmp][i];
				if(!visited[b]) que.push(b);
			}
		}
		cout<<endl;
	}

};

int main(){
	
	int vertex,edge;
	cout<<"input vertex number"<<endl;
	cin>>vertex;
	cout<<"input edge number"<<endl;
	cin>>edge;
	SparseGraph myGraph(vertex,false);
	for(int i = 0; i < edge; i++){
		int from,to;
		cout<<"add an edge between: "<<endl;
		cin>>from>>to;
		myGraph.addEdge(from,to);
	}
	myGraph.print();
	myGraph.dfs(0);
	myGraph.bfs(0);
    return 0;
}

File: javascript/sparseGraph.js
//this file is used to create a graph
//and realize breadth first search
//as well as depth first search

//initialize an adjacent matrix
//0 means no edge between these two vertexes
//1 means there is an edge from row - col
//we consider here a directed graph
let vertex = 5;
let matrix = new Array(vertex);
for(let i = 0; i < vertex; i++){
	matrix[i] = new Array(vertex);
	matrix[i].fill(0);
}

//add edge between u and v
//where 0 <= u,v <= vertex-1 && u != v
function addEdge(u,v){
	matrix[u][v] = 1;//change edge state to 1
	console.log("add an edge between given vertex");
	console.log(u);
	console.log(v);
	console.log("u to v is: ",matrix[u][v]);
}

function bfs(matrix){
	console.log("breadth first search result: ");
	let visited = new Array(vertex);
	visited.fill(false);
	let s = new Array();
	s.push(0);//start from the number 0 vertex
	while(s.length > 0){
		let tmp = s.pop();
		if(visited[tmp] == false){
			console.log(tmp);
		}
		visited[tmp] = true;
		for(let i = 0; i < vertex; i++){
			let layer = matrix[tmp][i];
			if(visited[i] == false && layer == 1){
				s.push(i);
			}
		}
	}
}

function dfs(matrix){
	console.log("depth first search result: ");
	let visited = new Array(vertex);
	visited.fill(false);
	let s = new Array();
	s.push(0);//start from the number 0 vertex
	while(s.length > 0){
		let tmp = s.shift();
		if(visited[tmp] == false){
			console.log(tmp);
		}
		visited[tmp] = true;
		for(let i = 0; i < vertex; i++){
			let layer = matrix[tmp][i];
			if(visited[i] == false && layer == 1){
				s.push(i);
			}
		}
	}
}

//here we have 5 vertexes, 0,1,2,3,4
addEdge(0,1);
addEdge(0,4);
addEdge(1,3);
addEdge(3,2);
addEdge(2,1);
dfs(matrix);
bfs(matrix);